lib/routes.js

```diff
@@ -145,29 +148,85 @@ module.exports = async function (app, opts) {
     verifyClient,
     lruGatewayResolvers,
     entityResolversFactory,
-    onlyPersisted,
-    persistedQueries = {},
+    persistedQuerySettings,
     allowBatchedQueries
   } = opts
 
+  // Load the persisted query settings
+  const {
+    isPersistedQuery,
+    getHash,
+    getQueryFromHash,
+    getHashForQuery,
+    saveQuery,
+    notFoundError,
+    notSupportedError
+  } = persistedQuerySettings || {}
+
+  const enablePersistedQueries = !!persistedQuerySettings
+
   async function execute (body, request, reply) {
     let { query } = body
-    const { persisted, operationName, variables } = body
+    const { operationName, variables } = body
+
+    // Generate the context for this request
+    let context = {}
+    if (contextFn) {
+      context = await contextFn(request, reply)
+    }
+
+    // Verify if a query matches the persisted format, if required
+    let persisted
+    if (enablePersistedQueries) {
+      persisted = isPersistedQuery(body)
+      if (persisted) {
+        // This is a peristed query, so we use the hash in the request
+        // to load the full query string.
+
+        // Extract the hash from the request
+        const hash = getHash && getHash(body)
+        if (hash) {
+          // Load the query for the provided hash
+          query = getQueryFromHash && await getQueryFromHash(hash)
+
+          if (!query) {
+            // Query has not been found, tell the client
+            throw new BadRequest(notFoundError)
+          }
 
-    if (persisted || onlyPersisted) {
-      query = persistedQueries[query]
+          // The query has now been set to the full query string
+        } else {
+          // This client should stop sending persisted queries,
+          // as we do not recognise them
+          throw new BadRequest(notSupportedError)
+        }
+      }
     }
 
+    // Validate a query is present
     if (!query) {
       return new BadRequest('Unknown query')
     }
 
-    let context = {}
-    if (contextFn) {
-      context = await contextFn(request, reply)
+    // Handle the query, throwing an error if required
+    const result = await reply.graphql(query, { pubsub: subscriber, ...context, __currentQuery: query }, variables, operationName)
+
+    // Check for not yet persisted queries
+    if (enablePersistedQueries && !persisted) {
+      // If provided the getHashForQuery, saveQuery settings we save this query
+      const hash = getHashForQuery && getHashForQuery(query)
```

Comment:
We have already computed the hash, so we should reuse what we computed before.

Reply:
The process is broken down into two steps:
1. Retrieve the hash from the graphql request (i.e. accessing the appropriate property of the request object), and loading the appropriate query from the given hash.
2. If the query is not (yet) persisted (i.e. 1 has failed), compute the hash from the query string, and save it.

The part you've commented on is step 2, where we compute the hash. We don't have a computed hash at this stage, and we have been unable to find a valid hash in the query (as per the `!persisted` check).